### ISSUE TITLE
FN single_match_else

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#3489]
`clippy 0.0.212 (b2601be 2018-11-27)`

I had this code which did not trigger any lints:
````rust
    let input = match directory {
        Some(value) => value,
        None => {
            return Err((
                ErrorKind::RemoveDirNoArg,
                "No argument assigned to --remove-dir, example: 'git-repos,registry-sources'"
                    .to_string(),
            ))
        }
    };
````

After applying rustfmt, an additional `;` was added:
````rust
    let input = match directory {
        Some(value) => value,
        None => {
            return Err((
                ErrorKind::RemoveDirNoArg,
                "No argument assigned to --remove-dir, example: 'git-repos,registry-sources'"
                    .to_string(),
            ));                                   // <--------- HERE
        }
    };
````

which caused clippy::single_match_else to trigger.

````
warning: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`
   --> src/library.rs:432:17
    |
432 |       let input = match directory {
    |  _________________^
433 | |         Some(value) => value,
434 | |         None => {
435 | |             return Err((
...   |
440 | |         }
441 | |     };
    | |_____^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match_else
help: try this
    |
432 |     let input = if let Some(value) = directory { value } else {
433 |     return Err((
434 |         ErrorKind::RemoveDirNoArg,
435 |         "No argument assigned to --remove-dir, example: 'git-repos,registry-sources'"
436 |             .to_string(),
437 |     ));
  ...
````

I think the link should also trigger on the unformatted example!

**Comment by mikerite** [comment]
I also prefer using `ok_or_else` and `?` to the suggestion.

```
    let input = directory.ok_or_else(|| Err(
        ErrorKind::RemoveDirNoArg,
        "No argument assigned to --remove-dir, example: 'git-repos,registry-sources'".to_string(),
    ))?;
```

**Comment by mikerite** [comment]
Looking at this again
```
    match Some(1) => {
        Some(_) => {...}
        None => { return 1 },
    }
```
is equivalent to
```
    match Some(1) => {
        Some(_) => {...}
        None => return 1,
    }
```
which is a common pattern and shouldn't be linted IMO.

If anything is changed, I think clippy shouldn't lint for the second pattern either.

**Comment by montrivo** [comment]
This lint currently doesn't lint cases where the "else" part of the 2 arm match is a single expression. The bug lies in that `{ return }` is considered as a single expression where as `{ return; }` isn't. 

One approach would be to consider single statement blocks equivalently. That would join @mikerite's last comment. That is, the three following examples wouldn't be linted. 

```rust
    match Some(1) => {
        Some(_) => {...}
        None => return 1,
    }
```

```rust
    match Some(1) => {
        Some(_) => {...}
        None => { return 1 },
    }
```
// this case is currently linted
```rust
    match Some(1) => {
        Some(_) => {...}
        None => { return 1; },
    }
```
